lib/encryptor.rb

```diff
@@ -46,30 +49,52 @@ def decrypt(*args, &block)
-      cipher = OpenSSL::Cipher::Cipher.new(options[:algorithm])
-      cipher.send(cipher_method)
-      if options[:iv]
-        cipher.iv = options[:iv]
-        if options[:salt].nil?
-          # Use a non-salted cipher.
-          # This behaviour is retained for backwards compatibility. This mode
-          # is not secure and new deployments should use the :salt options
-          # wherever possible.
-          cipher.key = options[:key]
-        else
-          # Use an explicit salt (which can be persisted into a database on a
-          # per-column basis, for example). This is the preferred (and more
-          # secure) mode of operation.
-          cipher.key = OpenSSL::PKCS5.pbkdf2_hmac_sha1(options[:key], options[:salt], 2000, cipher.key_len)
-        end
+  def crypt(cipher_method, *args) #:nodoc:
+    options = default_options.merge(:value => args.first).merge(args.last.is_a?(Hash) ? args.last : {})
+    raise ArgumentError.new('must specify a :key') if options[:key].to_s.empty?
+    cipher = OpenSSL::Cipher::Cipher.new(options[:algorithm])
+    cipher.send(cipher_method)
+    if options[:iv]
+      cipher.iv = options[:iv]
+      if options[:salt].nil?
+        # Use a non-salted cipher.
+        # This behaviour is retained for backwards compatibility. This mode
+        # is not secure and new deployments should use the :salt options
+        # wherever possible.
+        cipher.key = options[:key]
       else
-        cipher.pkcs5_keyivgen(options[:key])
+        # Use an explicit salt (which can be persisted into a database on a
+        # per-column basis, for example). This is the preferred (and more
+        # secure) mode of operation.
+        cipher.key = OpenSSL::PKCS5.pbkdf2_hmac_sha1(options[:key], options[:salt], 2000, cipher.key_len)
       end
-      yield cipher, options if block_given?
-      result = cipher.update(options[:value])
-      result << cipher.final
+    else
+      cipher.pkcs5_keyivgen(options[:key])
     end
+
+    yield cipher, options if block_given?
+
+    if options[:hmac_key]
+      digest   = OpenSSL::Digest.new(options[:hmac_algorithm])
+      hmac_key = options[:hmac_key]
+      if cipher_method == :decrypt
+        # perform an integrity check
+        secret_hash = options[:value][0.. digest.length-1]
+        cipher_text = options[:value][digest.length..-1]
+        # hash the cipher-text using :hmac_algorithm
+        integrity_check = OpenSSL::HMAC.digest(digest, hmac_key, cipher_text)
+        raise OpenSSL::HMACError.new('integrity check failed') if secret_hash != integrity_check
+        result = cipher.update(cipher_text) + cipher.final
+      else
+        # create hmac integrity check
+        result = cipher.update(options[:value]) + cipher.final
+        integrity_check = OpenSSL::HMAC.digest(digest, hmac_key, result)
+        result = integrity_check + result
+      end
+    else
+      result = cipher.update(options[:value]) + cipher.final
```

Comment:
nit: the previous code used `<<` which is mutating (thus faster) than the concatenation you're doing here